compiler/space.c

```diff
@@ -68,10 +68,21 @@ extern void report_b(FILE * out, const symbol * p) {
     }
 }
 
+extern void report_b_utf8(FILE * out, const symbol * p) {
+    int i;
+    for (i = 0; i < SIZE(p); i++) {
+        putc(p[i] & 0x0ff, out);
+    }
+}
```

Comment:
I don't think this does what you want it to - the name suggests it converts to UTF-8, but it actually takes each Unicode codepoint and just truncates it to fit a byte, ignoring all the upper bits - for example "ġ" is U+0121 so would get written as "!" (which is`\x21`.)

Reply:
In fact it is correct because the strings that are written by the Ada code generator are UTF-8.

The root issue is that the `char` is signed and when I populate the string buffer with some UTF-8 code
sequence the `0x80` and above sequences are converted to `0xFFxx` when the character is inserted
in the buffer. Hence, the test made by `report_b` will fail.  The `char` to `symbol` conversion occurs in
`str_append_ch` (the value I'm passing has the upper byte cleared by a `& 0x0FF`)

```C
extern void str_append_ch(struct str * str, char add) {

    symbol q[1];
    q[0] = add; // <<< conversion of signed character to symbol sets the upper byte
    str->data = add_to_b(str->data, 1, q);
}
```

I can come with another fix by adding the following functions:

```C
extern void str_append_symbol(struct str * str, symbol add) {

    str->data = add_to_b(str->data, 1, &add);
}

extern void write_symbol(struct generator * g, symbol s) {
    str_append_symbol(g->outbuf, s);
}
```

Then I'll use that new `write_symbol` in the Ada generator when writing the literal string.

```C
static void write_literal_string(struct generator * g, symbol * p) {
    int i;
    // Ada supports UTF-8 literal strings, we only need to escape the quote and
    // special characters.
    write_char(g, '"');
    for (i = 0; i < SIZE(p); i++) {
        int ch = p[i];
        if (ch == '"') {
            write_string(g, "\"\"");
        } else if (32 <= ch && ch < 127) {
            write_char(g, ch);
        } else if (ch <= 255) {
            write_symbol(g, ch & 0x0ff);
        } else {
            printf("In write_literal_string, can't convert p[%d] to char because it's 0x%02x\n", i, (int)p[i]);
            exit(1);
        }
    }
    write_char(g, '"');
}
```

Note, the sequence that I receive in `write_literal_string` is an UTF-8 sequence, I get 0xD0, 0xB2 which corresponds to в.

The test `if (ch < 255)` will catch incorrect literal strings.

Reply:
This seems to be a bug in `write_char()` to me - on platforms where `char` is signed by default, it currently only allows 0-127, but on platforms where `char` is unsigned by default (ARM is probably the most common one) 0-255 are allowed.

Previously it's been latent as the existing generators all write some form of string escape for characters 127 and higher.

I've pushed a fix for this to git master.

Does Ada allow arbitrary control characters in literal strings?  Characters < 32 and 127 will get written as literals by the above code.  It's a bit of a technicality since none of the Snowball stemmers we current ship use such characters, but the language does allow them.

Reply:
http://archive.adaic.com/standards/83lrm/html/lrm-02-06.html seems to say we can't directly put such control characters in a string literal:

> A string literal is formed by a sequence of graphic characters

and

> Similarly catenation of constants declared in the package ASCII can be used to obtain sequences of character values that include nongraphic character values (the so-called control characters). Examples of such uses of catenation are given below:
[...]
> "sequence that includes the" & ASCII.ACK & "control character" 

Quotes from http://archive.adaic.com/standards/83lrm/html/lrm-02-06.html (the equivalent page for Ada 95 seems to say the same, but doesn't mention the package `ASCII`.)

I'm going to push a change to give an error for such characters - they aren't expected to be needed in the domains Snowball targets, but if encountered it seems better to give a compile time error rather than generate invalid code.  If somebody with Ada knowledge wants to make this work then please send a patch.